### ISSUE TITLE
phone app internationalisation

### ISSUE THREAD

**Issue by PatGendre** [e-mission/e-mission-docs#354]
Hi, currently the messages in the phone app are hardcoded in the methods, have you considered extracting them out in text files with an i18n mechanism (e.g. a cordova lib) so that the app could be easily translated?

**Comment by shankari** [comment]
Yes, they should be extracted out. I had even asked the Paraguay project (that was in Spanish, and had funding) to do this change and contribute, but they ran out of money, so they only made changes to the javascript and didn't even do it in a reusable way.

Since both you (@PatGendre) and @jf87 presumably need this, is this something that you can contribute?

All of the notifications generated by the app should come through the unified logger plugin (in particular https://github.com/e-mission/cordova-unified-logger/blob/master/src/android/NotificationHelper.java), so you should be able to fix the methods and then easily find and replace their usages.

**Comment by shankari** [comment]
There is actually a project that will generate the translation text for us since we are open source.
https://translatewiki.net/wiki/Translating:New_project

We just need to set up the native code and the javascript to read the text from the translation files instead of hardcoding.

For the native code, this should be trivial since all visible notifications are generated from one class.
For the HTML, the Spanish PR above should have an indication of which strings need to be changed.
Here's an outline of how to do the HTML/JS translations right (which is also where I got the link to TranslateWiki from)
https://mclear.co.uk/2014/02/14/translations-language-i10n-i18n-cordova-phonegap-app/

**Comment by shankari** [comment]
If you do decide to take this on, I would suggest putting the translations into a separate repository since

> This requires a language addition procedure and commit access **(with direct push)** for at least one translatewiki.net staff member, to ensure that the work of translators will actually be integrated. 

**emphasis mine**

**Comment by PatGendre** [comment]
Ok, thanks for the indications! translatewiki.com seems a good idea (even if we will translate the texts ourselves); the [i18n method uses code from 2014](https://github.com/marcelklehr/html10n.js) so this has to be checked out. We'll have to do this translation so when we do it we will also take the extract from texts to files, but I cannot tell you when (before summer, I hope).

**Comment by shankari** [comment]
In the HTML code, all error popups should use the method
https://github.com/e-mission/e-mission-phone/blob/master/www/js/plugin/logger.js#L8

But strings that are not error popups may occur elsewhere. Should generally be in HTML only (`www/templates`) but there may be isolated instances in javascript as well.

**Comment by DO3B** [comment]
Hi, @PatGendre asked me to work on the internationalisation of your plugins. I had to change two plugins to do that : 

- e-mission-datacollection
- cordova-usercache

You will find enclosed these modifications. As you asked, I tried to respect the advices of Translate Wiki when naming the keys. For Android,  I used XML files to define the key-value associations. If we want to translate it in french, we will just need to create a folder `values-fr`, create associated files and translate the values. As for iOS, I added  a `en.lproj` with a `Localizable.strings` which will contains the key-value associations and called them by using `NSLocalizedStringFromTable(key, table, comment)` where `table` is the prefix of the `.strings` file of the plugin.

[e-mission-datacollection.zip](https://github.com/e-mission/e-mission-docs/files/3266060/e-mission-datacollection.zip)
[cordova-usercache.zip](https://github.com/e-mission/e-mission-docs/files/3266063/cordova-usercache.zip)

**Comment by shankari** [comment]
@DO3B thank you so much for this update, and I apologize for my late review.

tl;dr

1. Please submit PRs for what you have currently
2. We should move the translation files into their own repo after that

Details:

In terms of committing this, I think you can just submit a pull request to the plugins since they are also open source. Please make sure that you modify plugin.xml to copy the translated files to the correct location.

In a second step, it would be good to move the translations into their own repository, which I will create. translatewiki would like to have push access to the repository so that they can ensure that their changes are integrated in a timely fashion, and I don't want to give them push access to the repository code.
https://translatewiki.net/wiki/Translating:New_project

You would then also need to write a hook or otherwise copy the the translation files from the newly created repository into the project, so I would suggest making that change in a separate step.

**Comment by shankari** [comment]
@DO3B and this was the first (easy) step. The much bigger initiative will be to translate the phone UI.
Almost all of that should be in the HTML but there might be some text in the javascript as well.
This is the prior attempt at translating the UX, which should give you a pretty good idea of what needs to be changed.
https://github.com/e-mission/e-mission-phone/pull/431

**Comment by DO3B** [comment]
Hi @shankari,

As you asked, I created the two PRs.  Could you please create the TranslateWiki repository ? For the HTML/JS, I intend to use [angular-translate](https://angular-translate.github.io/ ).

**Comment by shankari** [comment]
@DO3B thank you for the quick turnaround. The repository for translations is at 
https://github.com/e-mission/e-mission-translate

angular-translate looks like a fine choice. I will plan on reaching out to the TranslateWiki team at the end of the week.

**Comment by shankari** [comment]
@DO3B just to ensure that we are on the same page, you are going to submit a PR to move the translation files into https://github.com/e-mission/e-mission-translate and then add hooks to copy them from there?

Maybe I should first reach out to the translate wiki team to see if they will accept submitting PRs into the existing repos as long as I commit to merging them quickly. Then we won't have to write hooks to copy things around.

Thoughts?

**Comment by DO3B** [comment]
It might be interesting to declare  **e-mission-translate** as a git submodule of **e-mission-phone** repo and still using a hook after the build. By doing that we ensure a full access to the translate repo to the Translate Wiki team and to update the translation we will just need to (theoretically) `git submodule update` and rebuild the app.

**Comment by shankari** [comment]
@DO3B that is pretty much what I had in mind. But then I saw some posts about problems with `git submodule` - e.g. https://github.blog/2016-02-01-working-with-submodules/, https://www.atlassian.com/blog/git/alternatives-to-git-submodule-git-subtree

You may want to look into submodules a bit more and see if they are suitable for our nee
ds, which are pretty modest, IMHO.

Note that we would have to set up the translation repo as a submodule of not just the e-mission-phone repo, but also the two native code repos.

**Comment by DO3B** [comment]
Indeed, it seems to be a bit too much for our needs and with a lot of constraints. As the hooks would be only in **e-mission-phone**, I thought that we needed a git submodule only once. For the moment, it might be sufficient to add instructions in the docs such as "Clone the git and copy/past the files in the `locales/` directory".

I propose to use the following architecture for the repo : 

- _language_/
values-_language_/
_language_-lproj/
_language_.json

and add two hooks (one for iOS and the other for Android) into **e-mission-phone** to check the locales available in a specific folder (i.e `locales/`) and to copy them into the right folders. I will try to do that this week :)

**Comment by shankari** [comment]
> Clone the git and copy/past the files in the locales/ directory

That is fine, but you can also add a hook that clones the repository (e.g. assume that there is a `git` executable and that you can execute`git clone` in a shell). You can then provide the manual instructions for troubleshooting in case somebody is using git without the command line?!

**Comment by shankari** [comment]
oops, I saw that you proposed that as well

> and add two hooks (one for iOS and the other for Android) into e-mission-phone to check the locales available in a specific folder (i.e locales/) and to copy them into the right folders. I will try to do that this week :)

Yup that sounds good! Hope you have a chance to work on it when it is so hot outside!?

**Comment by DO3B** [comment]
Hi, I created a PR (https://github.com/e-mission/e-mission-translate/pull/1) with the french translation (except consent). I also worked on the two hooks ([Android](https://github.com/fabmob/e-mission-phone-fabmob/blob/master/hooks/before_build/android/android_move_locales.js) and [iOS](https://github.com/fabmob/e-mission-phone-fabmob/blob/master/hooks/after_platform_add/ios/ios_move_locales.js)). I will try to work on the doc during the week.

**Comment by jf87** [comment]
This looks very cool @DO3B @shankari 
Do you think it's already save to create the same mapping for German translation?
Then I will start the process here :-)

**Comment by shankari** [comment]
@DO3B great! I am almost certainly going to merge the french translation without any comments 😃 
I will also reach out to TranslateWiki to start registering this as a project.

But I'm confused; if we are supporting internationalization fully, shouldn't the english translation be in https://github.com/e-mission/e-mission-translate as well?

**Comment by shankari** [comment]
And are you going to submit the two hooks as a PR as well?

**Comment by shankari** [comment]
@jf87 I think we will be fully done once @DO3B submits the hooks and the docs. But you can start working on the translation if it will take time...

**Comment by jf87** [comment]
Ok. I am just wondering if the "mapping doc" will change dramatically. If not, I can already give @DO3B current version to the person that will translate it.

But actually as you wrote, maybe there should now also be an English version so that the strings are not hard-coded anymore as well in that case.

**Comment by shankari** [comment]
@jf87 I don't think the "mapping doc" will change dramatically. You can give @DO3B's current version to the person that will translate it.

**Comment by shankari** [comment]
@DO3B I just remembered that there are a few additional files that need to be translated. I think you can deal with them after finishing the current round.

These are the files for the UX that configures the main e-mission plugins. Since they are so closely tied to the plugin, I checked them in to the plugin repos (e.g. https://github.com/e-mission/e-mission-data-collection/tree/master/www/ui/ionic). They are automatically copied to the native app as part of `plugin.xml` (e.g. https://github.com/e-mission/e-mission-data-collection/blob/master/plugin.xml#L16) In the case where the plugins are already added (e.g. for the devapp), they are downloaded by the 
[`bin/download_settings_controls.js`](https://github.com/e-mission/e-mission-phone/blob/master/bin/download_settings_controls.js) script.

**Comment by DO3B** [comment]
@shankari  As English is the fallback language of the project, we decided to leave the english files in **e-mission-phone** just in case if someone does not want i18n and just want to test if it works. Before submitting the two hooks, I need to make a third hook who will download the content of the **e-mission-translate** repo and put it in the `locales` folder.

**Comment by shankari** [comment]
@DO3B let's see what the translate wiki thinks about this structure. Also, once we have the hooks, the i18n will come as part of the build, so there are no additional steps when testing whether it works, right?

One potential enhancement is to maybe allow users to pick the locales that they want. I know that groups in developing countries are concerned about app size, and as the number of languages grows, it might be useful to allow groups that know that the dominant language(s) in their area to pick a subset of languages to copy - e.g. in Paraguay @myaman17 may want to only deploy in Spanish.

Then, using the same mechanism, people who don't want i18n would deploy only with english and not encounter any additional requirements.

We could put the list of languages into `config.xml` and have it default to english, for example.

**Comment by shankari** [comment]
@DO3B when I was testing my recent PRs, I saw an error like 

> pascalprecht.translate.$translateSanitization: No sanitization strategy has been configured. This can have serious security implications. See http://angular-translate.github.io/docs/#/guide/19_security for details.

Should we configure a sanitization strategy? `escape` seems pretty safe.

**Comment by DO3B** [comment]
The `escape` strategy seems good but the main issue will be that we won't be able to use html tag in our translation like in :

```json
"post-trip-map-display-tour-incident": "Zoom in as much as possible to the location where the incident occurred and click on the blue line of the trip to mark a <font size='+3'>&#x263B;</font> or <font size='+3'>&#x2639;</font> incident",
```

**Comment by shankari** [comment]
I think that is the only place where we use a HTML tag. Maybe we can change it to the text "smiley" or "frownie" instead until they get the sanitize to work better. Although looking at the issue, it doesn't look like it is going to be resolved any time soon.

**Comment by DO3B** [comment]
> @DO3B I just remembered that there are a few additional files that need to be translated. I think you can deal with them after finishing the current round.
> 
> These are the files for the UX that configures the main e-mission plugins. Since they are so closely tied to the plugin, I checked them in to the plugin repos (e.g. https://github.com/e-mission/e-mission-data-collection/tree/master/www/ui/ionic) - open to feedback on this.
> 
> They are automatically copied to the native app as part of `plugin.xml` (e.g. https://github.com/e-mission/e-mission-data-collection/blob/master/plugin.xml#L16) In the case where the plugins are already added (e.g. for the devapp), they are downloaded by the
> [`bin/download_settings_controls.js`](https://github.com/e-mission/e-mission-phone/blob/master/bin/download_settings_controls.js) script.
> 
> They are only displayed as part of the Developer Zone, so this is not critical, but it would be OK to get it done for completeness.

I am currently working on an "improved" version of the **Profile** screen. @PatGendre thinks that showing the current state as an indicator for the user could be useful to debug or just have a better understanding on how the phone app works. I currently need to translate those states and I think it might be better to parse the states directly in the phone app to be able to easily change the UI without having to modify plugins.
